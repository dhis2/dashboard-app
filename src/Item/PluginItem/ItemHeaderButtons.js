import React, { Component, Fragment } from 'react';
import SvgIcon from 'd2-ui/lib/svg-icon/SvgIcon';
import { REPORT_TABLE, CHART } from '../../util';

const style = {
    title: {
        paddingRight: 10,
        borderRight: '1px solid #ddd',
    },
};

const getIconButton = ({
    icon,
    targetType,
    onSelectVisualization,
    style = {},
}) => (
    <div
        style={{ cursor: 'pointer' }}
        onClick={() =>
            onSelectVisualization ? onSelectVisualization(targetType) : null
        }
    >
        <SvgIcon icon={icon} style={style} />
    </div>
);

class PluginItemHeaderButtons extends Component {
    render() {
        const { onSelectVisualization, onInterpretationsClick } = this.props;

        return (
            <Fragment>
                <div style={style.title}>
                    <div
                        style={{ cursor: 'pointer' }}
                        onClick={onInterpretationsClick}
                    >
                        <SvgIcon icon="Message" />
                    </div>
                </div>
                <div style={{ paddingLeft: 10, marginRight: 4 }}>
                    {getIconButton({
                        icon: 'ViewList',
                        targetType: REPORT_TABLE,
                        onSelectVisualization,
                    })}
                </div>
                <div style={{ marginRight: 4 }}>
                    {getIconButton({
                        icon: 'InsertChart',
                        targetType: CHART,
                        onSelectVisualization,
                    })}
                </div>
                {getIconButton({
                    icon: 'Public',
                })}
            </Fragment>
        );
    }
}

export default PluginItemHeaderButtons;
