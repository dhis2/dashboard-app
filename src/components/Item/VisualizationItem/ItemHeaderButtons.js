import React, { Component, Fragment } from 'react';
import MessageIcon from '@material-ui/icons/Message';
import TableIcon from '@material-ui/icons/ViewList';
import ChartIcon from '@material-ui/icons/InsertChart';
import MapIcon from '@material-ui/icons/Public';

import { extractFavorite } from './plugin';
import ItemHeaderButton from '../ItemHeaderButton';
import {
    VISUALIZATION_TYPE_TABLE,
    VISUALIZATION_TYPE_CHART,
    VISUALIZATION_TYPE_MAP,
    itemTypeMap,
    CHART,
    MAP,
    REPORT_TABLE,
    EVENT_CHART,
    EVENT_REPORT,
    DOMAIN_TYPE_AGGREGATE,
} from '../../../modules/itemTypes';
import { colors } from '../../../modules/colors';

const style = {
    iconBase: {
        width: '24px',
        height: '24px',
        fill: colors.lightMediumGrey,
    },
    buttonBase: {
        padding: '5px 6px 3px 6px',
    },
    buttonDisabled: {
        padding: '5px 6px 3px 6px',
        opacity: 0.5,
        cursor: 'unset',
    },
    toggleFooterPadding: {
        padding: '7px 6px 1px 6px',
    },
    border: {
        borderRadius: '2px',
        border: `1px solid ${colors.lightGrey}`,
    },
};

const baseStyle = {
    icon: style.iconBase,
    container: style.buttonBase,
};

const disabledStyle = {
    icon: style.iconBase,
    container: style.buttonDisabled,
};

const activeStyle = {
    icon: { ...style.iconBase, fill: colors.royalBlue },
    container: {
        ...style.buttonBase,
        backgroundColor: colors.lightBlue,
    },
};

const inactiveStyle = disabled => (disabled ? disabledStyle : baseStyle);

const tableBtnStyle = (activeType, disabled) =>
    [REPORT_TABLE, EVENT_REPORT].includes(activeType)
        ? activeStyle
        : inactiveStyle(disabled);

const chartBtnStyle = (activeType, disabled) =>
    [CHART, EVENT_CHART].includes(activeType)
        ? activeStyle
        : inactiveStyle(disabled);

const mapBtnStyle = (activeType, disabled) =>
    [MAP].includes(activeType) ? activeStyle : inactiveStyle(disabled);

export const getItemTypeId = (itemTypeMap, visualizationType, domainType) => {
    const item = Object.values(itemTypeMap).find(
        item =>
            item.visualizationType === visualizationType &&
            item.domainType === domainType
    );
    return item.id;
};

class VisualizationItemHeaderButtons extends Component {
    render() {
        const {
            item,
            onSelectVisualization,
            activeFooter,
            activeType,
            onToggleFooter,
        } = this.props;

        const domainType = itemTypeMap[item.type].domainType;

        const onViewTable = () =>
            onSelectVisualization(
                getItemTypeId(itemTypeMap, VISUALIZATION_TYPE_TABLE, domainType)
            );

        const onViewChart = () =>
            onSelectVisualization(
                getItemTypeId(itemTypeMap, VISUALIZATION_TYPE_CHART, domainType)
            );

        const onViewMap = () =>
            onSelectVisualization(
                getItemTypeId(itemTypeMap, VISUALIZATION_TYPE_MAP, domainType)
            );

        const toggleFooterBase = activeFooter ? activeStyle : baseStyle;

        const toggleFooter = {
            ...toggleFooterBase,
            container: {
                ...toggleFooterBase.container,
                ...style.toggleFooterPadding,
                ...style.border,
            },
        };

        // disable toggle buttons
        let disabled = false;

        if (item.type === VISUALIZATION_TYPE_CHART) {
            if (extractFavorite(item).type.match(/^YEAR_OVER_YEAR/)) {
                disabled = true;
            }
        }

        const tableButtonStyle = tableBtnStyle(activeType, disabled);
        const chartButtonStyle = chartBtnStyle(activeType, disabled);
        const mapButtonStyle = mapBtnStyle(activeType, disabled);

        return (
            <Fragment>
                <div style={{ marginRight: 10 }}>
                    <ItemHeaderButton
                        style={toggleFooter.container}
                        onClick={onToggleFooter}
                    >
                        <MessageIcon style={toggleFooter.icon} />
                    </ItemHeaderButton>
                </div>
                <div style={style.border}>
                    <ItemHeaderButton
                        disabled={disabled}
                        style={tableButtonStyle.container}
                        onClick={onViewTable}
                    >
                        <TableIcon style={tableButtonStyle.icon} />
                    </ItemHeaderButton>
                    <ItemHeaderButton
                        disabled={disabled}
                        style={chartButtonStyle.container}
                        onClick={onViewChart}
                    >
                        <ChartIcon style={chartButtonStyle.icon} />
                    </ItemHeaderButton>
                    {domainType === DOMAIN_TYPE_AGGREGATE ? (
                        <ItemHeaderButton
                            disabled={disabled}
                            style={mapButtonStyle.container}
                            onClick={onViewMap}
                        >
                            <MapIcon style={mapButtonStyle.icon} />
                        </ItemHeaderButton>
                    ) : null}
                </div>
            </Fragment>
        );
    }
}

export default VisualizationItemHeaderButtons;
