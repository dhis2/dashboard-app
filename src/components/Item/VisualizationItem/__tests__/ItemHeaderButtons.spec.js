import React from 'react';
import { shallow } from 'enzyme';
import ItemHeaderButtons from '../ItemHeaderButtons';

it('renders correctly', () => {
    const buttons = shallow(
        <ItemHeaderButtons
            item={{
                type: 'CHART',
                chart: { type: 'NOT_YOY', domainType: 'AGGREGATE' },
            }}
            visualization={{
                type: 'SINGLE_VALUE',
            }}
            onSelectVisualization={Function.prototype}
            activeFooter={false}
            activeType={'CHART'}
            onToggleFooter={Function.prototype}
        >
            My Little Pony
        </ItemHeaderButtons>
    );
    expect(buttons).toMatchSnapshot();
});
