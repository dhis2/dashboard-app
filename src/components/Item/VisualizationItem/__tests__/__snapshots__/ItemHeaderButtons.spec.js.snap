// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <VisualizationItemHeaderButtons
    activeFooter={false}
    activeType="CHART"
    item={
      Object {
        "chart": Object {
          "domainType": "AGGREGATE",
          "type": "NOT_YOY",
        },
        "type": "CHART",
      }
    }
    onSelectVisualization={[Function]}
    onToggleFooter={[Function]}
    visualization={
      Object {
        "type": "SINGLE_VALUE",
      }
    }
  >
    My Little Pony
  </VisualizationItemHeaderButtons>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <React.Fragment>
          <ItemHeaderButton
            onClick={[Function]}
            style={
              Object {
                "border": "1px solid #f3f5f7",
                "borderRadius": "2px",
                "padding": "7px 6px 1px 6px",
              }
            }
          >
            <pure(MessageIcon)
              style={
                Object {
                  "fill": "#a0adba",
                  "height": "24px",
                  "width": "24px",
                }
              }
            />
          </ItemHeaderButton>
        </React.Fragment>,
        null,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <ItemHeaderButton
            onClick={[Function]}
            style={
              Object {
                "border": "1px solid #f3f5f7",
                "borderRadius": "2px",
                "padding": "7px 6px 1px 6px",
              }
            }
          >
            <pure(MessageIcon)
              style={
                Object {
                  "fill": "#a0adba",
                  "height": "24px",
                  "width": "24px",
                }
              }
            />
          </ItemHeaderButton>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <pure(MessageIcon)
              style={
                Object {
                  "fill": "#a0adba",
                  "height": "24px",
                  "width": "24px",
                }
              }
            />,
            "onClick": [Function],
            "style": Object {
              "border": "1px solid #f3f5f7",
              "borderRadius": "2px",
              "padding": "7px 6px 1px 6px",
            },
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "style": Object {
                "fill": "#a0adba",
                "height": "24px",
                "width": "24px",
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
        "type": Symbol(react.fragment),
      },
      null,
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <React.Fragment>
            <ItemHeaderButton
              onClick={[Function]}
              style={
                Object {
                  "border": "1px solid #f3f5f7",
                  "borderRadius": "2px",
                  "padding": "7px 6px 1px 6px",
                }
              }
            >
              <pure(MessageIcon)
                style={
                  Object {
                    "fill": "#a0adba",
                    "height": "24px",
                    "width": "24px",
                  }
                }
              />
            </ItemHeaderButton>
          </React.Fragment>,
          null,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <ItemHeaderButton
              onClick={[Function]}
              style={
                Object {
                  "border": "1px solid #f3f5f7",
                  "borderRadius": "2px",
                  "padding": "7px 6px 1px 6px",
                }
              }
            >
              <pure(MessageIcon)
                style={
                  Object {
                    "fill": "#a0adba",
                    "height": "24px",
                    "width": "24px",
                  }
                }
              />
            </ItemHeaderButton>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": <pure(MessageIcon)
                style={
                  Object {
                    "fill": "#a0adba",
                    "height": "24px",
                    "width": "24px",
                  }
                }
              />,
              "onClick": [Function],
              "style": Object {
                "border": "1px solid #f3f5f7",
                "borderRadius": "2px",
                "padding": "7px 6px 1px 6px",
              },
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "style": Object {
                  "fill": "#a0adba",
                  "height": "24px",
                  "width": "24px",
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
          "type": Symbol(react.fragment),
        },
        null,
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
